python/ray/tune/function_runner.py

```diff
@@ -40,101 +49,212 @@ def __call__(self, **kwargs):
-            self.config.get("script_min_iter_time_s",
-                            self._default_config["script_min_iter_time_s"]))
-        result = self._status_reporter._get_and_clear_status()
-        while result is None:
-            time.sleep(1)
-            result = self._status_reporter._get_and_clear_status()
-
-        curr_ts_total = result.get(TIMESTEPS_TOTAL)
-        if curr_ts_total is not None:
-            result.update(
-                timesteps_this_iter=(
-                    curr_ts_total - self._last_reported_timestep))
-            self._last_reported_timestep = curr_ts_total
+        if not self._started:
+            # if not started, start
+            self._status_reporter._start()
+            self._runner.start()
+            self._started = True
+        elif self._runner.is_alive():
+            # if started and alive, inform the reporter to continue and
+            # generate the next result
+            self._continue_semaphore.release()
+
+        result = None
+        while result is None and self._runner.is_alive():
+            # fetch the next produced result
+            try:
+                result = self._results_queue.get(block=True,
+                                                 timeout=RESULT_FETCH_TIMEOUT)
+            except queue.Empty:
+                pass
+
+        # if no result were found, then the runner must no longer be alive
+        if result is None:
+            # Try one last time to fetch results in case results were reported
+            # in between the time of the last check and the termination of the
+            # thread runner.
+            try:
+                result = self._results_queue.get(block=False)
+            except queue.Empty:
+                pass
+
+        # check if error occured inside the thread runner
+        if result is None:
+            # only raise an error from the runner if all results are consumed
+            self._report_thread_runner_error(block=True)
+
+            # Under normal conditions, this code should never be reached since
+            # this branch should only be visited if the runner thread raised
+            # an exception. If no exception were raised, it means that the
+            # runner thread never reported any results which should not be
+            # possible when wrapping functions with
+            # `tune.trainable.wrap_function`.
+            raise TuneError((
+                    "Wrapped function ran until completion without reporting "
+                    "results or raising an exception."))
+
+        else:
+            if not self._error_queue.empty():
+                logger.warning((
+                        "Runner error waiting to be raised in main thread. "
+                        "Logging all available results first."
+                        ))
 
         return result
 
     def _stop(self):
-        self._status_reporter._stop()
+        # notify the status reporter to stop if still running
```

Comment:
in the spirit of simplicity, do we actually need any of this?

Reply:
Good catch! I added that before I had completely committed to synchronizing the `FunctionRunner` and the thread. I ended up keeping it as an undocumented way for the function to catch the stop (with the `should_stop` method I had).

Given this isn't really part of the API and not really supported, the whole thing should definitely get stripped.